truffle/src/main/java/org/jruby/truffle/core/regexp/RegexpNodes.java

```diff
@@ -446,24 +468,83 @@ public int hash(DynamicObject regexp) {
     @CoreMethod(names = "match_start", required = 2)
     public abstract static class MatchStartNode extends CoreMethodArrayArgumentsNode {
 
+        @Child private CallDispatchHeadNode dupNode;
         @Child private RopeNodes.MakeSubstringNode makeSubstringNode;
 
         public MatchStartNode(RubyContext context, SourceSection sourceSection) {
             super(context, sourceSection);
+            dupNode = DispatchHeadNodeFactory.createMethodCall(context);
             makeSubstringNode = RopeNodesFactory.MakeSubstringNodeGen.create(null, null, null);
         }
 
         @Specialization(guards = "isRubyString(string)")
-        public Object matchStart(DynamicObject regexp, DynamicObject string, int startPos) {
-            final Object matchResult = matchCommon(getContext(), this, makeSubstringNode, regexp, string, false, false, startPos);
+        public Object matchStart(VirtualFrame frame, DynamicObject regexp, DynamicObject string, int startPos) {
+            final DynamicObject dupedString = (DynamicObject) dupNode.call(frame, string, "dup");
+            final Object matchResult = matchCommon(getContext(), this, makeSubstringNode, regexp, dupedString, false, startPos);
+
             if (RubyGuards.isRubyMatchData(matchResult) && Layouts.MATCH_DATA.getRegion((DynamicObject) matchResult).numRegs > 0
                 && Layouts.MATCH_DATA.getRegion((DynamicObject) matchResult).beg[0] == startPos) {
                 return matchResult;
             }
+
             return nil();
         }
     }
 
+    @CoreMethod(names = "last_match", onSingleton = true, optional = 1, lowerFixnum = 1)
+    public abstract static class LastMatchNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        public Object lastMatch(NotProvided index) {
+            return getMatchData();
+        }
+
+        @Specialization
+        public Object lastMatch(VirtualFrame frame, int index,
+                                @Cached("create()") MatchDataNodes.GetIndexNode getIndexNode) {
+            final Object matchData = getMatchData();
+
+            if (matchData == nil()) {
+                return nil();
+            } else {
+                return getIndexNode.executeGetIndex(frame, matchData, index, NotProvided.INSTANCE);
+            }
+        }
+
+        @TruffleBoundary
+        private Object getMatchData() {
+            Frame frame = getContext().getCallStack().getCallerFrameIgnoringSend().getFrame(FrameInstance.FrameAccess.READ_WRITE, true);
```

Comment:
READ_WRITE access is fairly expensive, so maybe we should optimize by trying to access it in READ_ONLY first.
But here it seems the frame is never written to.